### ISSUE TITLE
Rethink TS penalty and LN-close adapter calibration

### ISSUE THREAD

**Issue by s3d-i** [ensomi-labs/ensomi-model#11]
## Problem

The current mapper combines a learned LN-close adapter with a fixed decode-time time-shift (TS) penalty. This works as a practical control knob, but the current formulation mixes quantities with different semantics and makes it hard to tell whether density/LN behavior is actually learned or imposed by decoding policy.

This issue tracks four related problems, with particular emphasis on the semantics of TS tokens.

### 1. The coefficients live in different units and act at different places

Current knobs include, for example:

- `lambda_ln_close`: weights the auxiliary LN-close loss during training;
- `close_scale`: converts LN-close logits into an additive vocabulary bias;
- `time_shift_length_penalty_alpha`: subtracts a fixed penalty from TS logits at inference.

These numbers are not directly comparable. One controls gradient magnitude, one controls an additive learned bias, and one changes the inference policy without participating in training. Treating them as if they were interchangeable scalar strengths obscures what each mechanism is actually doing.

### 2. The base token head and the LN-close adapter are not identifiable from token CE alone

The final token distribution sees roughly

```text
base_logits + state_prior_bias + ln_close_bias + grammar_mask
```

so token CE only constrains their sum. The base head can partially cancel or absorb the adapter, and the adapter can partially do work that the base head could have learned. The auxiliary LN-close loss helps supervise the head semantically, but it does not by itself determine the correct magnitude of its direct logit contribution.

We therefore need to distinguish:

- whether the close head predicts useful close hazard/state information;
- whether directly injecting that prediction into token logits improves constrained rollout;
- what calibration, if any, that injection should use.

### 3. A decode-only TS penalty creates a train/inference policy mismatch

The model is trained under one constrained token distribution and then decoded under another after subtracting a manually chosen TS penalty. A large improvement from this penalty proves that the raw policy is miscalibrated with respect to TS/event competition; it does **not** prove that the model has learned the desired density or musical structure.

Any production decode policy that materially changes action preferences should either be represented in the training objective or be justified as an explicit constrained inference procedure with separately measured trade-offs.

### 4. The TS penalty is semantically misaligned with the quantity we actually care about

This is the most important issue.

A TS token is not an ordinary semantic event. It represents elapsed time / continued absence of a chart event. Under the current sparse tokenizer, token index and musical time are different coordinates:

- several lane-action tokens can share the same `current_ms`;
- one TS token can advance seconds of real time;
- the same or nearly the same elapsed time can require different numbers of TS tokens because deltas are decomposed into vocabulary denominations.

For example, depending on the TS vocabulary, a 2000 ms gap may be represented by one token while a nearby 1990 ms gap may require several tokens. A fixed per-token penalty therefore does not correspond cleanly to either elapsed time or event density. It can also make two almost identical temporal gaps receive very different cumulative penalties.

More generally, TS and lane actions play different causal roles:

- TS answers **when the next event occurs**;
- lane actions answer **what happens at that event**.

Putting both in one flat categorical competition makes a fixed TS logit penalty an indirect and tokenizer-dependent way to control density. It may force more notes without improving beat alignment, motif structure, or playability.

## Possible architectural direction: factor time and mark decisions

A cleaner long-term formulation is to factor the next event as

```text
p(next timepoint, lane actions | history, audio, state)
  = p(delta_t | history, audio, state)
  * p(lane actions | delta_t, history, audio, state)
```

This would separate:

1. an **outer temporal process** that predicts the next event time / time delta;
2. an **inner mark decoder** that predicts the coupled lane actions at that timepoint under replay/grammar state.

Possible implementations include:

- discrete-time hazard / survival likelihood on a 10 ms or beat-relative grid;
- a learned delta distribution;
- a timepoint-level outer decoder plus a small autoregressive inner decoder for lane actions.

This is conceptually similar to compound-token / nested decoding approaches in symbolic music: preserve one outer step per semantic timepoint while still modeling dependencies among coupled features within the timepoint.

The main benefit is not just shorter sequences. It restores a useful causal separation:

```text
when does an event happen?
        ↓
what structured event happens there?
```

Then density is primarily a property of the learned temporal/event process rather than a side effect of penalizing TS vocabulary items.

## Near-term alternatives without rewriting the mapper

Before changing tokenization or the main architecture, test three more principled alternatives to a fixed TS penalty.

### A. Treat LN close as a calibrated hazard/auxiliary prediction, not an arbitrary fixed-strength bias

Make replay state (`open_mask`, `open_start_ms`, `open_age_ms`, etc.) available directly to the decoder hidden state so the core model can reason about LN state. Keep the LN-close head as an explicitly supervised hazard predictor.

Run ablations:

1. base model + replay-state embedding;
2. + auxiliary LN-close loss only;
3. + learned/calibrated direct close bias.

Do not assume `close_scale` should be fixed. If direct logit injection helps, fit or learn its calibration on held-out data and verify that it improves constrained rollout rather than only close-head accuracy.

### B. Replace fixed TS penalty with an explicit density constraint and learned dual variable

Define an expected onset/event count under the legal-token distribution and constrain it toward the control target instead of subtracting a hand-tuned constant from every TS token.

Conceptually:

```text
minimize token/structure loss
subject to |expected_density - target_density| <= epsilon
```

Use a primal-dual/Lagrangian update so the multiplier grows when the model violates the density constraint and shrinks when the constraint is already satisfied. This makes the strength of the correction data-dependent rather than a fixed `5.2`-style constant.

A first implementation can use teacher-forced legal states to estimate expected onset mass per time bin before attempting rollout-level differentiable objectives.

### C. Conditional unlikelihood for pathological skips, rather than globally penalizing every TS token

Do not make all TS tokens negative examples. Instead, penalize a TS choice only when it is conditionally undesirable, e.g. when:

- the local/remaining event budget is already substantially below target;
- selecting a long skip would worsen the deficit;
- the audio/control context strongly supports an onset.

This keeps valid musical silence possible while explicitly training against the failure mode we actually care about: pathological skipping/empty windows.

## Diagnostics required before choosing a replacement

Add legal-set diagnostics so we evaluate the policy the product actually uses, rather than only full-vocabulary raw logits:

- legal-set size and forced-step rate;
- target rank among legal tokens;
- probability mass on TS vs onset vs HOLD_END among legal actions;
- LN-close head calibration conditioned on open age/state;
- expected onset mass per real-time bin;
- sensitivity to audio/control/difficulty interventions;
- rollout density, alignment, longest gap, LN duration distribution, and pattern quality.

Also log gradient norms / cosine similarities for token CE, LN-close auxiliary loss, and density objectives before adopting automatic loss balancing. If the issue is gradient-scale imbalance, methods such as GradNorm are relevant; if objectives conflict, gradient-surgery methods are relevant. Neither fixes an incorrectly specified objective.

## Decision criteria

The goal is not simply to recover note count. A replacement should reduce reliance on hand-tuned decode-time TS penalties while preserving or improving:

- event density calibration across difficulty;
- beat/audio alignment;
- LN duration/closure behavior;
- empty-window / longest-gap rates;
- pattern diversity and structural quality;
- stability under constrained incremental rollout.

If the near-term approaches still require a large global TS penalty, that is evidence that the flat TS/event factorization itself should be reconsidered rather than tuned further.
